src/SingleLinkedList.h

```diff
@@ -222,62 +284,101 @@ T* SingleLinkedList<T>::remove(T* element) {
 	}
 }
 
+/*
+* Removes the element at a given index from the list.
+* index- position in list of the element to be removed. 
+* Throws exception if index in invalid.
+*/
 template <class T>
 T* SingleLinkedList<T>::remove(int index) {
-	cout << "Removed Element at index: " << index << endl;
-	return NULL;
+	if(index < 0 || index >= size) {
+		throw "ERROR: invalid index.";
+	} else if (index == 0){
+		return removeFirst();
+	} else if (index == size-1){
+		return removeLast();
+	} else {
+		Node<T>* currentNode = head;
+		int currentIndex = 0;
+		while(currentIndex != index-1){
+			currentNode = currentNode->next;
+			currentIndex++;
+		}
+		Node<T>* placeHolder = currentNode->next;
+		currentNode->next = currentNode->next->next;
+		size--;
+		T* element = placeHolder->data;
+		delete placeHolder;
+		delete currentNode;
+		return element;
+	}
 }
 
 /* Set method */
+/*
+*Sets the element of the given index to the given element.
+* index- index of element to be set.
+* element- value or object to be set. 
+* Throws exception if index is invalid.
+*/ 
 
 template <class T>
 void SingleLinkedList<T>::set(int index, T* element) {
 	cout << "Set Element (" << *element << ") at index: " << index << endl;
+	if(index < 0 || index >= size) {
+		throw "ERROR: Invalid index.";
+	}
+	Node<T>* node = head;
+	for(int i = 0; i < index ; i++) {
+		node = node->next;
+	}
+	node->data = element;
+	delete node;
```

Comment:
I think we need the following code here:

```c++
delete (node->data)
node->data = element
```

Double check me though.

Reply:
correct me if i'm wrong, looks like that would free up the memory the node's data points to, then assigns the element to what was just free'd (that node's data value). we'd still need to delete the node.
wouldn't just reassigning what the node's data points to, then deleting the node basically do the same thing?
or am i missing something...

Reply:
So when we are setting something we erase the location of the old object that was stored in the node. However both objects still exist within the heap. So to compensate for garbage collection we delete the nodes data that we are overwriting and reassign the pointer to this new data we are setting.

If we delete the node we will end up breaking the list cause we are setting the previous nodes next to null in essence. So if we had a list of a->b->c->d->null and we call set(2, F) what ends up happening is a->b->null d->null. So now we can get from b to d. And what was set at index two is also no longer accessible. Not only that but both the data set and overwritten in the node c are no longer accessible as well so we cause memory leakage.

Hopefully that makes sense.